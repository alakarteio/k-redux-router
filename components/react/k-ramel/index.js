!function(e,t){"object"==typeof exports&&"undefined"!=typeof module?t(exports,require("@k-redux-router/core"),require("k-ramel"),require("react"),require("prop-types"),require("@k-ramel/react")):"function"==typeof define&&define.amd?define(["exports","@k-redux-router/core","k-ramel","react","prop-types","@k-ramel/react"],t):t(e["@k-redux-router/react-kramel"]={},null,null,e.React,null,null)}(this,function(e,t,r,n,o,u){"use strict";var i="default"in t?t.default:t;n=n&&n.hasOwnProperty("default")?n.default:n,o=o&&o.hasOwnProperty("default")?o.default:o;var a=function(e,t){if(!(e instanceof t))throw new TypeError("Cannot call a class as a function")},c=function(){function e(e,t){for(var r=0;r<t.length;r++){var n=t[r];n.enumerable=n.enumerable||!1,n.configurable=!0,"value"in n&&(n.writable=!0),Object.defineProperty(e,n.key,n)}}return function(t,r,n){return r&&e(t.prototype,r),n&&e(t,n),t}}(),l=function(e,t,r){return t in e?Object.defineProperty(e,t,{value:r,enumerable:!0,configurable:!0,writable:!0}):e[t]=r,e},f=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var r=arguments[t];for(var n in r)Object.prototype.hasOwnProperty.call(r,n)&&(e[n]=r[n])}return e},s=function(e,t){var r={};for(var n in e)t.indexOf(n)>=0||Object.prototype.hasOwnProperty.call(e,n)&&(r[n]=e[n]);return r},p=function(e,t){if(!e)throw new ReferenceError("this hasn't been initialised - super() hasn't been called");return!t||"object"!=typeof t&&"function"!=typeof t?e:t},d=function(){return function(e,t){if(Array.isArray(e))return e;if(Symbol.iterator in Object(e))return function(e,t){var r=[],n=!0,o=!1,u=void 0;try{for(var i,a=e[Symbol.iterator]();!(n=(i=a.next()).done)&&(r.push(i.value),!t||r.length!==t);n=!0);}catch(e){o=!0,u=e}finally{try{!n&&a.return&&a.return()}finally{if(o)throw u}}return r}(e,t);throw new TypeError("Invalid attempt to destructure non-iterable instance")}}(),h=function(e,t){return function(r){var o,u;return u=o=function(o){function u(r,n){a(this,u);var o=p(this,(u.__proto__||Object.getPrototypeOf(u)).call(this,r,n));return o.toShow=function(){var r=o.context.store.drivers.router;if(r.getState()&&r.getResult()){var n=r.getCurrentRoute();if(t&&t.absolute){var u=e===n.code;u!==o.state.show&&o.setState(function(e){return f({},e,{show:u})})}else{for(var i=e===n.code;n&&n.parent&&!i;)n=r.getRoute(n.parent),i=e===n.code;i!==o.state.show&&o.setState(function(e){return f({},e,{show:i})})}}else console.error("[k-redux-router] | There is no route found in `state.ui.router.result`")},o.state={show:!1},o}return function(e,t){if("function"!=typeof t&&null!==t)throw new TypeError("Super expression must either be null or a function, not "+typeof t);e.prototype=Object.create(t&&t.prototype,{constructor:{value:e,enumerable:!1,writable:!0,configurable:!0}}),t&&(Object.setPrototypeOf?Object.setPrototypeOf(e,t):e.__proto__=t)}(u,n.Component),c(u,[{key:"componentWillMount",value:function(){var e=this;this.unsubscribe=this.context.store.subscribe(function(){e.toShow()}),this.toShow()}},{key:"componentWillUnmount",value:function(){this.unsubscribe()}},{key:"render",value:function(){return this.state.show?n.createElement(r,this.props):null}}]),u}(),o.displayName=function(e){return"router("+(e.displayName||e.name||e.constructor&&e.constructor.name||"Unknown")+")"}(r),o.contextTypes={store:function(){return null}},u}};h.absolute=function(e,t){return h(e,f({},t,{absolute:!0}))};var y=function(e){var t=e.href,r=e.onClick,o=s(e,["href","onClick"]),u=o.className,i=o.children,a="",c=t.base;return t.compiled&&(c=t.compiled(o)),o.query&&(a="?"+toQueryString(o.query)),c=""+c+a,n.createElement("a",{className:u,href:c,onClick:r},i)};y.propTypes={className:o.string,children:o.node,href:o.shape({base:o.string,compiled:o.func}),onClick:o.func},y.defaultProps={className:void 0,children:void 0,href:void 0,onClick:void 0};var v=u.inject(function(e,t,r){var n=t.onClick,o=t.code,u=t.query,i=s(t,["onClick","code","query"]),a=r.router;return{href:a.getRoute(o).href,onClick:function(e){n&&n(e),function(e){return function(e){return!!(e.shiftKey||e.altKey||e.metaKey||e.ctrlKey)}(e)||function(e){return e.button&&0!==e.button}(e)||e.defaultPrevented}(e)||(a.push(o,i,u),e.preventDefault())}}})(y);e.router=function(){var e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:{},n=e.routes,o=e.path,u=void 0===o?"ui.router":o,a=e.getState,c=void 0===a?function(e){return e.ui.router}:a,s=i(n,{getState:c}),p=s.middleware,h=s.reducer,y=s.init;return{getDriver:function(e){return f({},Object.entries(t.actions).reduce(function(t,r){var n=d(r,2),o=n[0],u=n[1];return f({},t,l({},o,function(){return e.dispatch(u.apply(void 0,arguments))}))},{}),Object.entries(t.selectors(c)).reduce(function(t,r){var n=d(r,2),o=n[0],u=n[1],i=function(){return u(e.getState())};return["getRoute","getPathParam","getQueryParam","getParam"].includes(o)&&(i=function(){return u.apply(void 0,arguments)(e.getState())}),f({},t,l({},o,i))},{}))},getReducer:function(){return{path:u,reducer:h}},getEnhancer:function(){return r.applyMiddleware(p)},init:function(e){e.dispatch(y())}}},e.forRoute=h,e.Link=v,Object.defineProperty(e,"__esModule",{value:!0})});
//# sourceMappingURL=index.js.map
