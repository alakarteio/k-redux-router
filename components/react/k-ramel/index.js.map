{"version":3,"file":"index.js","sources":["src/hoc.jsx","src/link/link.jsx","src/link/link.container.js","src/driver.js"],"sourcesContent":["import React from 'react'\n\n// TODO: factorize code with react-redux\nconst getDisplayName = Component => `router(${\n  Component.displayName\n  || Component.name\n  || (Component.constructor && Component.constructor.name)\n  || 'Unknown'\n})`\n\nconst isRouteFound = code => result => result && [].concat(code).includes(result.code)\n\nconst hoc = (code, options) => Component => class extends React.Component {\n  static displayName = getDisplayName(Component)\n\n  static contextTypes = {\n    store: () => null, // this is to avoid importing prop-types\n  }\n\n  constructor(props, context) {\n    super(props, context)\n\n    this.state = { show: false }\n  }\n\n  componentWillMount() {\n    // subscribe\n    this.unsubscribe = this.context.store.subscribe(() => {\n      this.toShow()\n    })\n\n    // run in once\n    this.toShow()\n  }\n\n  componentWillUnmount() {\n    this.unsubscribe()\n  }\n\n  toShow = () => {\n    const { router } = this.context.store.drivers\n\n    // TODO: hardcoded path (state.ui)\n    if (!router.getState() || !router.getResult()) {\n      // eslint-disable-next-line no-console\n      console.error('[k-redux-router] | There is no route found in `state.ui.router.result`')\n      return\n    }\n\n    // Absolute mode, we are looking in top level only\n    let currentRoute = router.getCurrentRoute()\n    if (options && options.absolute) {\n      const show = (code === currentRoute.code)\n      if (show !== this.state.show) {\n        this.setState(innerState => ({ ...innerState, show }))\n      }\n\n      return\n    }\n\n    // Either way we are looking top down the result tree\n    let show = (code === currentRoute.code)\n    while (currentRoute && currentRoute.parent && !show) {\n      currentRoute = router.getRoute(currentRoute.parent)\n      show = (code === currentRoute.code)\n    }\n\n    if (show !== this.state.show) {\n      this.setState(innerState => ({ ...innerState, show }))\n    }\n  }\n\n  render() {\n    if (!this.state.show) return null\n\n    return <Component {...this.props} />\n  }\n}\n\nhoc.absolute = (code, options) => hoc(code, { ...options, absolute: true })\n\nexport default hoc\n","import React from 'react'\nimport PropTypes from 'prop-types'\n\n// TODO: factorize code with react-redux\nconst Link = (props) => {\n  const {\n    href,\n    onClick,\n    ...params,\n  } = props\n\n  const {\n    className,\n    children\n  } = params\n\n  // TODO: factorise code from middleware and from here\n  let queryPart = ''\n  let processedHref = href.base\n  if (href.compiled) processedHref = href.compiled(params)\n  if (params.query) queryPart = `?${toQueryString(params.query)}`\n  processedHref = `${processedHref}${queryPart}`\n\n  return (\n    <a\n      className={className}\n      href={processedHref}\n      onClick={onClick}\n    >\n      {children}\n    </a>\n  )\n}\n\nLink.propTypes = {\n  className: PropTypes.string,\n  children: PropTypes.node,\n  href: PropTypes.shape({\n    base: PropTypes.string,\n    compiled: PropTypes.func,\n  }),\n  onClick: PropTypes.func,\n}\n\nLink.defaultProps = {\n  className: undefined,\n  children: undefined,\n  href: undefined,\n  onClick: undefined,\n}\n\nexport default Link\n","import { inject } from '@k-ramel/react'\nimport Component from './link'\n\n// copied from redux-little-router\n// - this part is about letting user make CTRL+CLICK (example)\n// - but keep the push-history-api available when we do a simple click\nconst LEFT_MOUSE_BUTTON = 0\nconst isNotLeftClick = e => e.button && e.button !== LEFT_MOUSE_BUTTON\nconst hasModifier = e => !!(e.shiftKey || e.altKey || e.metaKey || e.ctrlKey)\nconst shouldIgnoreClick = e => hasModifier(e) || isNotLeftClick(e) || e.defaultPrevented\n\n// map store TODO: use driver (since we are in k-ramel)\nconst mapStore = (store, { onClick, code, query, ...params }, { router }) => ({\n  href: router.getRoute(code).href,\n  onClick: (e) => {\n    // parent onClick callback\n    if (onClick) onClick(e)\n\n    // from redux-little-router\n    if (shouldIgnoreClick(e)) return\n\n    // dispatch the push\n    router.push(code, params, query)\n\n    // prevent default behaviour\n    e.preventDefault()\n  },\n})\n\nexport default inject(mapStore)(Component)\n","import factory, { selectors, actions } from '@k-redux-router/core'\nimport { applyMiddleware } from 'k-ramel'\n\nexport default (options = {}) => {\n  const {\n    routes,\n    path = 'ui.router',\n    getState = state => state.ui.router,\n  } = options\n\n  const { middleware, reducer, init } = factory(routes, { getState })\n\n  return {\n    getDriver: store => ({\n      // actions\n      ...Object\n        .entries(actions)\n        .reduce(\n          (acc, [key, action]) => ({\n            ...acc,\n            [key]: (...args) => store.dispatch(action(...args)),\n          }),\n          {},\n        ),\n      // selectors\n      ...Object.\n        entries(selectors(getState))\n        .reduce(\n          (acc, [key, selector]) => {\n            let fn = () => selector(store.getState())\n            if (['getRoute', 'getPathParam', 'getQueryParam' ,'getParam' ].includes(key)) {\n              fn = (...args) => selector(...args)(store.getState())\n            }\n\n            return ({\n              ...acc,\n              [key]: fn,\n            })\n          },\n          {},\n        )\n    }),\n    getReducer: () => ({ path, reducer }),\n    getEnhancer: () => applyMiddleware(middleware),\n    init: store => { store.dispatch(init()) },\n  }\n}\n"],"names":["hoc","code","options","props","context","toShow","router","_this","store","drivers","getState","getResult","currentRoute","getCurrentRoute","absolute","show","state","setState","innerState","parent","getRoute","error","React","Component","unsubscribe","this","subscribe","displayName","name","constructor","getDisplayName","contextTypes","Link","href","onClick","params","className","children","queryPart","processedHref","base","compiled","query","toQueryString","propTypes","PropTypes","string","node","shape","func","defaultProps","undefined","inject","e","shiftKey","altKey","metaKey","ctrlKey","hasModifier","button","isNotLeftClick","defaultPrevented","shouldIgnoreClick","push","preventDefault","routes","path","ui","factory","middleware","reducer","init","Object","entries","actions","reduce","acc","key","action","dispatch","selectors","selector","fn","includes","applyMiddleware"],"mappings":"m2DAYMA,EAAM,SAACC,EAAMC,UAAY,sDAOjBC,EAAOC,4EACXD,EAAOC,aAmBfC,OAAS,eACCC,EAAWC,EAAKH,QAAQI,MAAMC,QAA9BH,UAGHA,EAAOI,YAAeJ,EAAOK,iBAO9BC,EAAeN,EAAOO,qBACtBX,GAAWA,EAAQY,cACfC,EAAQd,IAASW,EAAaX,KAChCc,IAASR,EAAKS,MAAMD,QACjBE,SAAS,wBAAoBC,GAAYH,wBAO9CA,EAAQd,IAASW,EAAaX,KAC3BW,GAAgBA,EAAaO,SAAWJ,KAC9BT,EAAOc,SAASR,EAAaO,UACpClB,IAASW,EAAaX,KAG5Bc,IAASR,EAAKS,MAAMD,QACjBE,SAAS,wBAAoBC,GAAYH,yBAvBtCM,MAAM,6EAvBXL,OAAUD,MAAM,wUAViCO,EAAMC,2EAevDC,YAAcC,KAAKrB,QAAQI,MAAMkB,UAAU,aACzCrB,gBAIFA,6DAIAmB,sDAqCAC,KAAKT,MAAMD,KAETO,gBAACC,EAAcE,KAAKtB,OAFE,gBA5DxBwB,YAVc,6BACrBJ,EAAUI,aACPJ,EAAUK,MACTL,EAAUM,aAAeN,EAAUM,YAAYD,MAChD,eAMkBE,CAAeP,KAE7BQ,oBACE,kBAAM,WA+DjB/B,EAAIc,SAAW,SAACb,EAAMC,UAAYF,EAAIC,OAAWC,GAASY,UAAU,MC3EpE,IAAMkB,EAAO,SAAC7B,OAEV8B,EAGE9B,EAHF8B,KACAC,EAEE/B,EAFF+B,QACGC,IACDhC,sBAGFiC,EAEED,EAFFC,UACAC,EACEF,EADFE,SAIEC,EAAY,GACZC,EAAgBN,EAAKO,YACrBP,EAAKQ,WAAUF,EAAgBN,EAAKQ,SAASN,IAC7CA,EAAOO,QAAOJ,MAAgBK,cAAcR,EAAOO,aACpCH,EAAgBD,EAGjChB,+BACac,OACLG,UACGL,OAOfF,EAAKY,qBACQC,EAAUC,gBACXD,EAAUE,UACdF,EAAUG,YACRH,EAAUC,gBACND,EAAUI,eAEbJ,EAAUI,MAGrBjB,EAAKkB,6BACQC,gBACDA,YACJA,eACGA,GC1CX,MAuBeC,SAjBE,SAAC5C,WAAS0B,IAAAA,QAASjC,IAAAA,KAAMyC,IAAAA,MAAUP,kCAAY7B,IAAAA,mBACxDA,EAAOc,SAASnB,GAAMgC,aACnB,SAACoB,GAEJnB,GAASA,EAAQmB,GAPC,mBADN,qBAAQA,EAAEC,UAAYD,EAAEE,QAAUF,EAAEG,SAAWH,EAAEI,SACtCC,CAAYL,IAFpB,mBAAKA,EAAEM,QADJ,IACcN,EAAEM,OAEOC,CAAeP,IAAMA,EAAEQ,iBAUhEC,CAAkBT,OAGfU,KAAK9D,EAAMkC,EAAQO,KAGxBsB,qBAISZ,CAAiB7B,2BC1BhBrB,4DAEZ+D,EAGE/D,EAHF+D,SAGE/D,EAFFgE,KAAAA,aAAO,gBAELhE,EADFQ,SAAAA,aAAW,mBAASM,EAAMmD,GAAG7D,YAGO8D,EAAQH,GAAUvD,aAAhD2D,IAAAA,WAAYC,IAAAA,QAASC,IAAAA,sBAGhB,wBAENC,OACAC,QAAQC,WACRC,OACC,SAACC,kBAAMC,OAAKC,mBACPF,OACFC,EAAM,kBAAarE,EAAMuE,SAASD,mCAKtCN,OACDC,QAAQO,YAAUtE,IACjBiE,OACC,SAACC,kBAAMC,OAAKI,OACNC,EAAK,kBAAMD,EAASzE,EAAME,oBACzB,WAAY,eAAgB,gBAAiB,YAAayE,SAASN,OACjE,kBAAaI,yBAAAA,CAAkBzE,EAAME,mBAIvCkE,OACFC,EAAMK,sBAML,kBAAShB,OAAMI,wBACd,kBAAMc,kBAAgBf,SAC7B,cAAiBU,SAASR"}