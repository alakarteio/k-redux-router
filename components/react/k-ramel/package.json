{
  "name": "@k-redux-router/react-k-ramel",
  "version": "1.0.1",
  "description": "React components (binded to redux store with k-ramel)",
  "main": "dist/index.umd.js",
  "module": "dist/index.es.js",
  "source": "src/index.js",
  "author": "Fabien JUIF <fabien.juif@gmail.com>",
  "license": "MIT",
  "scripts": {
    "build:core:es": "cross-env NODE_ENV=build rollup -c ./misc/rollup.config.js",
    "build:core:umd": "cross-env NODE_ENV=build FORMAT=umd rollup -c ./misc/rollup.config.js",
    "build": "run-p build:*:*"
  },
  "devDependencies": {
    "@babel/cli": "^7.2.0",
    "@babel/core": "^7.2.2",
    "@babel/plugin-proposal-class-properties": "^7.2.1",
    "@babel/plugin-proposal-object-rest-spread": "^7.2.0",
    "@babel/preset-react": "^7.0.0",
    "cross-env": "^5.1.5",
    "npm-run-all": "^4.1.5",
    "react": "^16.6.3",
    "react-test-renderer": "^16.6.3",
    "rollup": "^0.68.0",
    "rollup-plugin-babel": "^4.1.0",
    "rollup-plugin-commonjs": "^9.2.0",
    "rollup-plugin-terser": "^3.0.0"
  },
  "peerDependencies": {
    "@k-ramel/react": "^1.2.0",
    "@k-redux-router/core": "^1.0.0",
    "k-ramel": "^1.0.0",
    "prop-types": "^15.6.0",
    "react": "^16.0.0"
  },
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org/"
  },
  "babel": {
    "presets": [
      "@babel/preset-react"
    ],
    "plugins": [
      "@babel/plugin-proposal-object-rest-spread",
      "@babel/plugin-proposal-class-properties"
    ]
  }
}
