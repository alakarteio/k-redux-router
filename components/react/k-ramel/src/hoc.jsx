import React from 'react'

// TODO: factorize code with react-redux
const getDisplayName = Component => `router(${
  Component.displayName
  || Component.name
  || (Component.constructor && Component.constructor.name)
  || 'Unknown'
})`

const hoc = (code, options) => Component => class extends React.Component {
  static displayName = getDisplayName(Component)

  static contextTypes = {
    store: () => null, // this is to avoid importing prop-types
  }

  constructor(props, context) {
    super(props, context)

    this.state = { show: false }
  }

  componentWillMount() {
    // subscribe
    this.unsubscribe = this.context.store.subscribe(() => {
      this.toShow()
    })

    // run in once
    this.toShow()
  }

  componentWillUnmount() {
    this.unsubscribe()
  }

  toShow = () => {
    const { router } = this.context.store.drivers

    // TODO: hardcoded path (state.ui)
    if (!router.getState() || !router.getResult()) {
      // eslint-disable-next-line no-console
      console.error('[k-redux-router] | There is no route found in `state.ui.router.result`')
      return
    }

    // Absolute mode, we are looking in top level only
    let currentRoute = router.getCurrentRoute()
    if (options && options.absolute) {
      const show = (code === currentRoute.code)
      if (show !== this.state.show) {
        this.setState(innerState => ({ ...innerState, show }))
      }

      return
    }

    // Either way we are looking top down the result tree
    let show = (code === currentRoute.code)
    while (currentRoute && currentRoute.parent && !show) {
      currentRoute = router.getRoute(currentRoute.parent)
      show = (code === currentRoute.code)
    }

    if (show !== this.state.show) {
      this.setState(innerState => ({ ...innerState, show }))
    }
  }

  render() {
    if (!this.state.show) return null

    return <Component {...this.props} />
  }
}

hoc.absolute = (code, options) => hoc(code, { ...options, absolute: true })

export default hoc
