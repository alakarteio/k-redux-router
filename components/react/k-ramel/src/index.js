export { default as driver } from './driver'
export { default as forRoute } from './hoc'
export { default as Link } from './link'
