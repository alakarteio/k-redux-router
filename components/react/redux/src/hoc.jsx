import React from 'react'
import { selectors } from '@k-redux-router/core'

const getDisplayName = Component => `router(${
  Component.displayName
  || Component.name
  || (Component.constructor && Component.constructor.name)
  || 'Unknown'
})`

const hoc = (code, options = {}) => (Component) => {
  const {
    getState = (state => state.ui.router),
    absolute = false,
  } = options

  const { getResult, getRoute, getCurrentRoute } = selectors(getState)

  return class extends React.Component {
    static displayName = getDisplayName(Component)

    static contextTypes = {
      store: () => null, // this is to avoid importing prop-types
    }

    constructor(props, context) {
      super(props, context)

      this.state = { show: false }
    }

    componentWillMount() {
      // subscribe
      this.unsubscribe = this.context.store.subscribe(() => {
        this.toShow()
      })

      // run in once
      this.toShow()
    }

    componentWillUnmount() {
      this.unsubscribe()
    }

    toShow = () => {
      const state = this.context.store.getState()

      if (!getState(state) || !getResult(state)) {
        // eslint-disable-next-line no-console
        console.error('[k-redux-router] | There is no route found in `state.ui.router.result`')
        return
      }

      // Absolute mode, we are looking in top level only
      let currentRoute = getCurrentRoute(state)
      if (absolute) {
        const show = (code === currentRoute.code)
        if (show !== this.state.show) {
          this.setState(innerState => ({ ...innerState, show }))
        }

        return
      }

      // Either way we are looking top down the result tree
      let show = (code === currentRoute.code)
      while (currentRoute && currentRoute.parent && !show) {
        currentRoute = getRoute(currentRoute.parent)(state)
        show = (code === currentRoute.code)
      }

      if (show !== this.state.show) {
        this.setState(innerState => ({ ...innerState, show }))
      }
    }

    render() {
      if (!this.state.show) return null

      return <Component {...this.props} />
    }
  }
}

hoc.absolute = (code, options) => hoc(code, { ...options, absolute: true })

export default hoc
