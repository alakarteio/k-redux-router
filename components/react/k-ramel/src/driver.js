import { router, selectors, actions } from '@k-redux-router/core'
import { applyMiddleware } from 'k-ramel'

export default (options = {}) => {
  const {
    routes,
    path = 'ui.router',
    getState = state => state.ui.router,
  } = options

  const { middleware, reducer, init } = router(routes, { getState })

  return {
    getDriver: store => ({
      // actions
      ...Object
        .entries(actions)
        .reduce(
          (acc, [key, action]) => ({
            ...acc,
            [key]: (...args) => store.dispatch(action(...args)),
          }),
          {},
        ),
      // selectors
      ...Object.entries(selectors(getState))
        .reduce(
          (acc, [key, selector]) => {
            let fn = () => selector(store.getState())
            if (['getRoute', 'getPathParam', 'getQueryParam', 'getParam'].includes(key)) {
              fn = (...args) => selector(...args)(store.getState())
            }

            return ({
              ...acc,
              [key]: fn,
            })
          },
          {},
        ),
    }),
    getReducer: () => ({ path, reducer }),
    getEnhancer: () => applyMiddleware(middleware),
    init: (store) => { store.dispatch(init()) },
  }
}
