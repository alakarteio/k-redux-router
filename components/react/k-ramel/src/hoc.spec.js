/* eslint-env jest */
import React from 'react'
import renderer from 'react-test-renderer'
import { createStore } from 'k-ramel'
import { provider } from '@k-ramel/react'
import driver from './driver'
import hoc from './hoc'

const matchApp = code => () => {
  const store = createStore(
    {
      dummy: (state = 'nothing') => state, // dummy reducer to get rid of error from redux
    },
    {
      drivers: {
        router: {
          getDriver: () => ({
            getState: () => 'state',
            getResult: () => 'result',
            getRoute: () => ({
              code: code.split('_')[0],
            }),
            getCurrentRoute: () => ({
              code, // current route is the root parameter
              parent: code.split('_')[0], // either relative or absolute or undefined
            })
          })
        }
      }
    }
  )

  const RelativeDecoratedComponent = hoc('relative')(() => <div>Relative component</div>)
  const AbsoluteDecoratedComponent = hoc.absolute('absolute')(() => <div>Absolute component</div>)
  const App = provider(store)(() => (
    <div>
      <RelativeDecoratedComponent />
      <AbsoluteDecoratedComponent />
    </div>
  ))

  expect(renderer.create(<App />)).toMatchSnapshot()
}

describe('hoc', () => {
  it('should print relative route -strict equality-', matchApp('relative'))
  it('should print relative route -child-', matchApp('relative_child'))
  it('should print absolute route -strict equality-', matchApp('absolute'))
  it('should not print absolute route -child-', matchApp('absolute_child'))
})
