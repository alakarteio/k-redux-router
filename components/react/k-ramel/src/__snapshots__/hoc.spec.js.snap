// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hoc array of codes should not print absolute route -child- 1`] = `<div />`;

exports[`hoc array of codes should print absolute route -strict equality- 1`] = `
<div>
  <div>
    Absolute component
  </div>
</div>
`;

exports[`hoc array of codes should print relative route -child- 1`] = `
<div>
  <div>
    Relative component
  </div>
</div>
`;

exports[`hoc array of codes should print relative route -strict equality- 1`] = `
<div>
  <div>
    Relative component
  </div>
</div>
`;

exports[`hoc not found should print the failover component when route is not found 1`] = `
<div>
  <div>
    Not found component
  </div>
</div>
`;

exports[`hoc uniq code (string) should not print absolute route -child- 1`] = `<div />`;

exports[`hoc uniq code (string) should print absolute route -strict equality- 1`] = `
<div>
  <div>
    Absolute component
  </div>
</div>
`;

exports[`hoc uniq code (string) should print relative route -child- 1`] = `
<div>
  <div>
    Relative component
  </div>
</div>
`;

exports[`hoc uniq code (string) should print relative route -strict equality- 1`] = `
<div>
  <div>
    Relative component
  </div>
</div>
`;
